File: 0464_can_i_win/solution.rb
# LeetCode 0464 - Can I Win
# https://leetcode.com/problems/can-i-win/

class Solution
  def can_i_win(max_choosable_integer, desired_total)
    return true if desired_total <= 0

    total = max_choosable_integer * (max_choosable_integer + 1) / 2
    return false if total < desired_total

    @memo = {}

    can_win?(0, 0, max_choosable_integer, desired_total)
  end

  alias_method :canIWin, :can_i_win

  private

  def can_win?(state, current_total, max_choosable_integer, desired_total)
    return @memo[state] if @memo.key?(state)

    (1..max_choosable_integer).each do |pick|
      bit = 1 << (pick - 1)
      next if (state & bit) != 0

      return @memo[state] = true if current_total + pick >= desired_total
      return @memo[state] = true unless can_win?(state | bit, current_total + pick, max_choosable_integer, desired_total)
    end

    @memo[state] = false
  end
end

File: 0464_can_i_win/solution.ts
// LeetCode 0464 - Can I Win
// https://leetcode.com/problems/can-i-win/

export class Solution {
    canIWin(maxChoosableInteger: number, desiredTotal: number): boolean {
        if (desiredTotal <= 0) return true;
        const total = (maxChoosableInteger * (maxChoosableInteger + 1)) / 2;
        if (total < desiredTotal) return false;

        const memo = new Map<number, boolean>();

        const canWin = (state: number, currentTotal: number): boolean => {
            if (memo.has(state)) return memo.get(state)!;
            for (let pick = 1; pick <= maxChoosableInteger; pick += 1) {
                const bit = 1 << (pick - 1);
                if (state & bit) continue;
                if (currentTotal + pick >= desiredTotal) {
                    memo.set(state, true);
                    return true;
                }
                if (!canWin(state | bit, currentTotal + pick)) {
                    memo.set(state, true);
                    return true;
                }
            }
            memo.set(state, false);
            return false;
        };

        return canWin(0, 0);
    }
}
